### PR TITLE
gf: init at unstable-2022-06-22

### PR DESCRIPTION
**Pull request by 0xd61**
###### Description

[gf](https://github.com/nakst/gf) is a simple GDB frontend with minimal dependencies.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by AndersonTorres** [comment]
It certainly does not fit contributing.md

File: maintainers/maintainer-list.nix [hunk 1]
```diff
@@ -85,6 +85,12 @@
     githubId = 56617252;
     matrix = "@oxc45:matrix.org";
   };
+  _0xd61 = {
```
Comment (AndersonTorres) [inline maintainers/maintainer-list.nix:88]: Put this in a separate commit: `maintainers: add 0xd61`

File: pkgs/development/tools/misc/gf/default.nix [hunk 2]
```diff
@@ -0,0 +1,66 @@
+{ lib, stdenv, makeWrapper
+
```
Comment (AndersonTorres) [inline pkgs/development/tools/misc/gf/default.nix:2]: One attribute per line.
A single space after comma - `, likeThis`

File: pkgs/development/tools/misc/gf/default.nix [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+  ,extensionFolder ? null
+}:
+
+let
+  revision = "e0d6d2f59344f853a4a204d5313db6b6a5e5de7d";
+in
+
+stdenv.mkDerivation rec {
+  pname = "gf";
+  version = "2+${builtins.substring 0 7 revision}";
```
Comment (AndersonTorres) [inline pkgs/development/tools/misc/gf/default.nix:15]: This is not Archlinux.

Put something like `version = unstable-YYYY-MM-DD;`

File: pkgs/development/tools/misc/gf/default.nix [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+  src = fetchgit {
+    url = "https://github.com/nakst/gf.git";
+    rev = "${revision}";
```
Comment (AndersonTorres) [inline pkgs/development/tools/misc/gf/default.nix:19]: Put the opaque revision directly here.

Better: use `fetchFromGitHub` instead of a crude `fetchgit`.

File: pkgs/development/tools/misc/gf/default.nix [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    sha256 = "1b9iggz50l9aksblg1lj64daf2l5kjnkixdw8qjjblp63xz668r7";
+  };
+
+  nativeBuildInputs = [ makeWrapper pkg-config ];
+  buildInputs = [ libX11 gdb ]
+    ++ lib.optional freetypeSupport freetype;
+
+  inherit extensionFolder;
```
Comment (AndersonTorres) [inline pkgs/development/tools/misc/gf/default.nix:27]: What is this??

File: pkgs/development/tools/misc/gf/default.nix [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+  postPatch = ''
+    if [ ! -z "$extensionFolder" ]; then
+      echo adding extensions from $extensionFolder;
+      cp -r "$extensionFolder"/* .
+    fi
+  '';
+
+  patchPhase = ''
+    echo #!/usr/bin/env bash > build.sh
```
Comment (AndersonTorres) [inline pkgs/development/tools/misc/gf/default.nix:37]: What are you trying to do here? This code is messy!

File: pkgs/development/tools/misc/gf/default.nix [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+  '';
+
+  installPhase = ''
+    mkdir -p "$out/bin"
+    cp gf2 "$out/bin/"
+    runHook postInstall
+  '';
+
+  postInstall = with lib; ''
+    wrapProgram $out/bin/gf2 --prefix PATH : ${makeBinPath[ gdb ]}
```
Comment (AndersonTorres) [inline pkgs/development/tools/misc/gf/default.nix:56]:
```suggestion
    wrapProgram $out/bin/gf2 --prefix PATH : ${lib.makeBinPath[ gdb ]}
```
```suggestion
  postInstall =  ''
    wrapProgram $out/bin/gf2 --prefix PATH : ${lib.makeBinPath[ gdb ]}
```

File: pkgs/development/tools/misc/gf/default.nix [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+    runHook postPatch
+  '';
+
+
+  buildPhase = ''
+    extra_flags=-DUI_FREETYPE_SUBPIXEL ./build.sh
+  '';
+
+  installPhase = ''
+    mkdir -p "$out/bin"
```
Comment (AndersonTorres) [inline pkgs/development/tools/misc/gf/default.nix:50]:
```suggestion
    runHook preInstall
    mkdir -p "$out/bin"
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/development/tools/misc/gf/default.nix [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+in
+
+stdenv.mkDerivation rec {
+  pname = "gf";
+  version = "2+${builtins.substring 0 7 revision}";
+
+  src = fetchgit {
+    url = "https://github.com/nakst/gf.git";
+    rev = "${revision}";
+    sha256 = "1b9iggz50l9aksblg1lj64daf2l5kjnkixdw8qjjblp63xz668r7";
```
Comment (viraptor) [inline pkgs/development/tools/misc/gf/default.nix:20]: I think you were working on a different cached version. This sha doesn't match the checkout.
```
error: hash mismatch in fixed-output derivation '/nix/store/a1sgn19j9w6k37ydf835il48vfiq7r05-gf-e0d6d2f.drv':
         specified: sha256-JyNjfh/m0iUlRrz1OK2chQqnGjGShkeXnipRUP57Ma0=
            got:    sha256-OCo84IgQlkvZrkpYj22ppTecwWXQBsVJxooFazmx1AU=
```

File: maintainers/maintainer-list.nix (hunk 1 above)
Comment (0xd61) [reply maintainers/maintainer-list.nix:88]: done

File: pkgs/development/tools/misc/gf/default.nix (hunk 2 above)
Comment (0xd61) [reply pkgs/development/tools/misc/gf/default.nix]: done

File: pkgs/development/tools/misc/gf/default.nix (hunk 3 above)
Comment (0xd61) [reply pkgs/development/tools/misc/gf/default.nix]: done

File: pkgs/development/tools/misc/gf/default.nix (hunk 4 above)
Comment (0xd61) [reply pkgs/development/tools/misc/gf/default.nix]: done

File: pkgs/development/tools/misc/gf/default.nix (hunk 5 above)
Comment (0xd61) [reply pkgs/development/tools/misc/gf/default.nix:28]: The gf author provides extensions for purchase. The files of this extensions folder must be copied into the source directory. I wanted to add an option to add this extension folder. I'm still new to nix and probably there is a better option for doing this, but it worked for me. It is similar to how ST handles patches ([link](https://github.com/NixOS/nixpkgs/blob/62de360d75419386b69ea4af4a8ce104949a8242/pkgs/applications/terminal-emulators/st/default.nix#L26)).

File: pkgs/development/tools/misc/gf/default.nix (hunk 6 above)
Comment (0xd61) [reply pkgs/development/tools/misc/gf/default.nix]: fixed

File: pkgs/development/tools/misc/gf/default.nix (hunk 7 above)
Comment (0xd61) [reply pkgs/development/tools/misc/gf/default.nix]: done

File: pkgs/development/tools/misc/gf/default.nix (hunk 8 above)
Comment (0xd61) [reply pkgs/development/tools/misc/gf/default.nix:50]: done

File: pkgs/development/tools/misc/gf/default.nix (hunk 9 above)
Comment (0xd61) [reply pkgs/development/tools/misc/gf/default.nix]: Should be fixed now.

**Comment by 0xd61** [comment]
@AndersonTorres and @viraptor thanks for the review and the suggested changes.

File: pkgs/development/tools/misc/gf/default.nix (hunk 5 above)
Comment (AndersonTorres) [reply pkgs/development/tools/misc/gf/default.nix:28]: Well, I took some time reading the README.
I need to know a thing or two.

First, those third party extensions are loaded when the program is running, or they need to be present on build time?

Because, if they are to be loaded at runtime, then you do not need to include them at all - you just tell the running program to read them somehow.
This is not a case similar to ST, because ST patches explicitly modify the source code.

Else, the things are a bit more complicated. 

I suppose those extensions are handled by a proprietary license, and they can't be easily downloaded. There are some support for such proprietary things on Nixpkgs, however this is not recommended.

File: pkgs/development/tools/misc/gf/default.nix (hunk 6 above)
Comment (AndersonTorres) [reply pkgs/development/tools/misc/gf/default.nix]: Well, I have read the original code, and it is so messy it would be easier to rewrite it instead of circumventing it.

File: pkgs/development/tools/misc/gf/default.nix (hunk 5 above)
Comment (0xd61) [reply pkgs/development/tools/misc/gf/default.nix:28]: The extensions are cpp files and need to be present during build time.

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/development/tools/misc/gf/default.nix (hunk 6 above)
Comment (0xd61) [reply pkgs/development/tools/misc/gf/default.nix]: I replaced the sed with a patchfile. Or did you mean I should rewite the whole build.sh file?

File: pkgs/development/tools/misc/gf/default.nix [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+    if [ ! -z "$extensionFolder" ]; then
+      echo adding extensions from $extensionFolder;
+      cp -r "$extensionFolder"/* .
+    fi
+  '';
+
+  patchPhase = ''
+    patchShebangs \
+    build.sh
+    sed -i 's|if \[ -d /usr/include/freetype2 \]; then extra_flags="$extra_flags -lfreetype -D UI_FREETYPE -I /usr/include/freetype2"; |if pkg-config --cflags freetype2 > /dev/null; then extra_flags="$extra_flags -D UI_FREETYPE `pkg-config --cflags --libs freetype2`"|g' build.sh
```
[the remark this answers is not in the document]
Comment (0xd61) [reply pkgs/development/tools/misc/gf/default.nix]: done

File: pkgs/development/tools/misc/gf/default.nix [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+  postPatch = ''
+    patchShebangs build.sh
+
+    if [ ! -z "$extensionFolder" ]; then
+      echo adding extensions from $extensionFolder;
+      cp -r "$extensionFolder"/* .
+    fi
+  '';
+
+  patches = ./build-use-optional-freetype-with-pkg-config.patch;
```
Comment (AndersonTorres) [inline pkgs/development/tools/misc/gf/default.nix:39]:
```suggestion
  patches = [
    ./build-use-optional-freetype-with-pkg-config.patch 
  ];
```

File: pkgs/development/tools/misc/gf/default.nix (hunk 5 above)
Comment (AndersonTorres) [reply pkgs/development/tools/misc/gf/default.nix:28]: Hum, that's interesting. It requires another approach.

I read the `build.sh` code, however I am not seeing how it detects and builds the extensions. Nothing like a `gcc *.cpp` on it.

Maybe the extension file also modifies the build script?

Until better judgement, remove all the extension stuff.

File: pkgs/development/tools/misc/gf/default.nix [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+  '';
+
+  installPhase = ''
+    runHook preInstall
+    mkdir -p "$out/bin"
+    cp gf2 "$out/bin/"
+    runHook postInstall
+  '';
+
+  postInstall = ''
```
Comment (AndersonTorres) [inline pkgs/development/tools/misc/gf/default.nix:52]:
```suggestion
  postFixup = ''
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/development/tools/misc/gf/default.nix (hunk 6 above)
Comment (AndersonTorres) [reply pkgs/development/tools/misc/gf/default.nix]: To be honest I prefer to hack the upstream sources, lol

It is OK for now.

File: pkgs/development/tools/misc/gf/default.nix (hunk 5 above)
Comment (0xd61) [reply pkgs/development/tools/misc/gf/default.nix:28]: The files are detected directly in the source. https://github.com/nakst/gf/blob/e0d6d2f59344f853a4a204d5313db6b6a5e5de7d/gf2.cpp#L1246-L1252

Comment (0xd61) [reply pkgs/development/tools/misc/gf/default.nix:28]: What is wrong with the current approach? Are there any side effects?

File: pkgs/development/tools/misc/gf/default.nix (hunk 11 above)
Comment (0xd61) [reply pkgs/development/tools/misc/gf/default.nix:39]: done

File: pkgs/development/tools/misc/gf/default.nix (hunk 12 above)
Comment (0xd61) [reply pkgs/development/tools/misc/gf/default.nix:52]: done

File: pkgs/development/tools/misc/gf/default.nix (hunk 5 above)
Comment (AndersonTorres) [reply pkgs/development/tools/misc/gf/default.nix:28]: Now I understood it better. It uses preprocessor magic to include the extensions. It has the same effect as a patch file.
Now I have a better idea.

File: pkgs/development/tools/misc/gf/default.nix [hunk 13]
```diff
@@ -0,0 +1,63 @@
+{ lib
+, stdenv
+, makeWrapper
+, fetchFromGitHub
+, libX11
+, pkg-config
+, gdb
+, freetype
+, freetypeSupport ? false
+, extensionFolder ? null
```
Comment (AndersonTorres) [inline pkgs/development/tools/misc/gf/default.nix:10]:
```suggestion
, extensions ? []
```

File: pkgs/development/tools/misc/gf/default.nix [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+    owner = "nakst";
+    rev = "e0d6d2f59344f853a4a204d5313db6b6a5e5de7d";
+    sha256 = "01fln4wnn1caqr4wa1nhcp0rqdx5m5nqyn2amvclp5hhi3h3qaiq";
+  };
+
+  nativeBuildInputs = [ makeWrapper pkg-config ];
+  buildInputs = [ libX11 gdb ]
+    ++ lib.optional freetypeSupport freetype;
+
+  inherit extensionFolder;
```
Comment (AndersonTorres) [inline pkgs/development/tools/misc/gf/default.nix:28]:
```suggestion
```

File: pkgs/development/tools/misc/gf/default.nix [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+
+  postPatch = ''
+    patchShebangs build.sh
+
+    if [ ! -z "$extensionFolder" ]; then
+      echo adding extensions from $extensionFolder;
+      cp -r "$extensionFolder"/* .
+    fi
+  '';
```
Comment (AndersonTorres) [inline pkgs/development/tools/misc/gf/default.nix:37]:
```suggestion
  postPatch = lib.forEach extensions (ext: ''
      cp ${ext} .
  )'';
```

File: pkgs/development/tools/misc/gf/default.nix [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+
```
Comment (AndersonTorres) [inline pkgs/development/tools/misc/gf/default.nix:38]:
```suggestion
   preConfigure = ''
     patchShebangs build.sh
   '';

```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/development/tools/misc/gf/default.nix (hunk 15 above)
Comment (0xd61) [reply pkgs/development/tools/misc/gf/default.nix:37]: Thanks for your suggestions. This helps a lot, however with this approach the filenames are prefixed with the nix-hash (e.g. 3p75yihvn1k56ax019i0grqvwg4fja2s-extensions.cpp). Is there a simple way to parse the original filename? Simply checking for the "-" will not work if the filename also has a dash in it. Otherwise the preprocessor magic does not recognize the files.

Comment (AndersonTorres) [reply pkgs/development/tools/misc/gf/default.nix:37]: Maybe `cp ${ext} ./${ext.name}` should help...

Comment (0xd61) [reply pkgs/development/tools/misc/gf/default.nix]: Awesome, thanks for your input. It looks much cleaner now.

File: pkgs/top-level/all-packages.nix [hunk 17]
```diff
@@ -16660,6 +16660,10 @@ with pkgs;
     readline = readline81;
   };
 
+  gf = callPackage ../development/tools/misc/gf {
+    freetypeSupport = true;
+  };
+
```
Comment (AndersonTorres) [inline pkgs/top-level/all-packages.nix:16666]:
```suggestion
  gf = callPackage ../development/tools/misc/gf { };

```

File: pkgs/development/tools/misc/gf/default.nix [hunk 18]
```diff
@@ -0,0 +1,62 @@
+{ lib
+, stdenv
+, makeWrapper
+, fetchFromGitHub
+, libX11
+, pkg-config
+, gdb
+, freetype
+, freetypeSupport ? false
```
Comment (AndersonTorres) [inline pkgs/development/tools/misc/gf/default.nix:9]:
```suggestion
, freetypeSupport ? true
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/top-level/all-packages.nix (hunk 17 above)
Comment (0xd61) [reply pkgs/top-level/all-packages.nix]: done

Comment (0xd61) [reply pkgs/top-level/all-packages.nix]: done

File: pkgs/development/tools/misc/gf/default.nix (hunk 18 above)
Comment (0xd61) [reply pkgs/development/tools/misc/gf/default.nix]: done

**Review by AndersonTorres — CHANGES_REQUESTED** [review]
Change the commit message:

`gf: init at unstable-2022-06-22`

File: pkgs/development/tools/misc/gf/default.nix [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+  postPatch = lib.forEach extensions (ext: ''
+      cp ${ext} ./${ext.name or (builtins.baseNameOf ext)}
+  '');
+
+   preConfigure = ''
+     patchShebangs build.sh
+   '';
+
+  buildPhase = ''
+    extra_flags=-DUI_FREETYPE_SUBPIXEL ./build.sh
```
Comment (AndersonTorres) [inline pkgs/development/tools/misc/gf/default.nix:41]: I almost forgot: decorate this phase with `runHook`s

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/development/tools/misc/gf/default.nix (hunk 19 above)
Comment (0xd61) [reply pkgs/development/tools/misc/gf/default.nix:42]: done

**Review by AndersonTorres — APPROVED** [review]
If no one complains in 2 days, ping me and I will merge.

**Comment by 0xd61** [comment]
@AndersonTorres  thanks for your help!
